app/daemon.go

```diff
@@ -300,51 +298,89 @@ func (d *MenderShellDaemon) gotAuthToken(p []dbus.SignalParams, needsReconnect b
 	return jwtToken
 }
 
-func (d *MenderShellDaemon) needsReconnect() bool {
-	select {
-	case e := <-d.reconnectChan:
-		log.Tracef("needsReconnect: got event: %s", e.event)
-		return true
-	case <-time.After(time.Second):
-		return false
-	}
-}
-
-func (d *MenderShellDaemon) dbusEventLoop(client mender.AuthClient) {
+func (d *MenderShellDaemon) eventLoop(client mender.AuthClient) {
+	var err error
 	needsReconnect := false
+	done := d.ctx.Done()
+	tokenChan := client.GetJwtTokenStateChangeChannel()
+EventLoop:
 	for {
-		if d.shouldStop() {
-			break
-		}
+		select {
+		case <-done:
+			break EventLoop
+
+		case p := <-tokenChan:
+			if len(p) > 0 && p[0].ParamType == dbus.GDBusTypeString {
+				token := d.gotAuthToken(p, needsReconnect)
+				if len(token) > 0 {
+					log.Tracef("eventLoop: got a token len=%d", len(token))
+					needsReconnect = false
+				}
+			}
 
-		if d.needsReconnect() {
-			log.Trace("dbusEventLoop: daemon needs to reconnect")
-			needsReconnect = true
-		}
+		case event := <-d.eventChan:
+			log.Tracef("eventLoop: got event: %s", event.event)
+			switch event.event {
+			case EventReconnectRequest:
+				log.Trace("eventLoop: daemon needs to reconnect")
+				if d.authorized {
+					jwtToken, serverURL, _ := client.GetJWTToken()
+					e := MenderShellDaemonEvent{
+						event: EventReconnect,
+						data:  jwtToken,
+						id:    "(eventLoop)",
+					}
+					log.Tracef("(eventLoop) posting Event: %s", e.event)
+					d.serverUrl = serverURL
+					d.postEvent(e)
+					needsReconnect = false
+				}
+			case EventReconnect:
+				err = connectionmanager.Reconnect(
+					ws.ProtoTypeShell, d.serverUrl,
+					d.deviceConnectUrl, event.data,
+					d.skipVerify, d.serverCertificate,
+					config.MaxReconnectAttempts, d.ctx,
+				)
+				if err != nil {
+					switch errors.Cause(err) {
+					case connectionmanager.ErrClientUnauthorized:
+						jwtToken, serverURL, _ := client.GetJWTToken()
+						if jwtToken != event.data {
+							// JWT changed, let's try again
+							// But let's clear any pending token-
+							// change events.
+							select {
+							case <-tokenChan:
+							default:
+							}
+							// JWT changed, let's try again
+							d.serverUrl = serverURL
+							err = connectionmanager.Reconnect(
+								ws.ProtoTypeShell, d.serverUrl,
+								d.deviceConnectUrl, event.data,
+								d.skipVerify, d.serverCertificate,
+								config.MaxReconnectAttempts, d.ctx,
+							)
+							if err == nil {
+								continue
```

Comment:
From here, we should still execute the `else` below, right?:
```
log.Trace("eventLoop: reconnected")
d.connectionEstChan <- MenderShellDaemonEvent{
    event: EventConnectionEstablished,
}
```

Reply:
True, thanks!